where-to-travel/src/main/webapp/script.js

```diff
@@ -1,212 +1,401 @@
 }
 
 /**
- * Populates map with pins. Given a list of places, put markers at each
- * lat/long location.
- *
- * @param {array} placeArray array of places to place markers at
+ * Obtains user's location from browser and sets home location. If error
+ * occurs, message is displayed to user in modal.
  */
-function populatePlaces(placeArray) {
-  let i;
-  for(i = 0; i < placeArray.length; i++) {
-    let name = placeArray[i].name;
-    let coordinates = placeArray[i].geometry.location;
-    let placeMarker = new google.maps.Marker({
-      position: coordinates,
-      map: map,
-      title: name
-    });
-
-    let infowindow = new google.maps.InfoWindow({
-      content: name
-    });
-
-    placeMarker.addListener('click', function() {
-      infowindow.open(map, placeMarker);
-    });
-  }
+function useLocation() {
+  getUserLocation().then(homeObject => {
+    home = homeObject;
+    setHomeMarker(); 
+  }).catch(message => {
+    const messageContent = '<p>' + message + '</p>'; 
+    openModal(messageContent);
+  });
 }
- 
 
-/** 
- * If browser supports geolocation and user provides permissions, obtains user's 
- * latitude and longitude. Otherwise, asks user to input address and converts input
- * to latitude and longitude.
+/**
+ * If browser supports geolocation and user provides location permissions, obtains user's
+ * latitude and longitude.
  *
- * @return {Object} Contains latitude and longitude corresponding to user's location
+ * @return {Promise} Fulfilled promise is object containing lat/lng and rejected promise
+ *                   is string message describing why obtaining the location failed.
  */
- function getUserLocation() {
+function getUserLocation() {
   return new Promise(function(resolve, reject) {
-  
     function success(position) {
-      return resolve({lat: position.coords.latitude, lng: position.coords.longitude});
+      resolve({
+        lat: position.coords.latitude,
+        lng: position.coords.longitude,
+      });
     }
 
-   function error() {
-     return resolve(getLocationFromUserInput());
-   }
+    function deniedAccessUserLocation() {
+      reject('Browser does not have permission to access location. ' +
+             'Please enable location permissions or enter an address to ' + 
+             'set a home location.');
+    }
 
     if (navigator.geolocation) {
-      navigator.geolocation.getCurrentPosition(success, error);
+      navigator.geolocation.getCurrentPosition(success, deniedAccessUserLocation);
     } else {
-      return resolve(getLocationFromUserInput());
+      reject('Browser does not support geolocation. Please enter an ' +
+             'address to set a home location.');
     }
   });
 }
 
+/**
+ * Reads address entered by user and passes to Geocoding API to get lat/lng
+ * and set home location. If error occurs, message is displayed to user in modal.
+ */
+function useAddress() {
+  const address = document.getElementById('addressInput').value;  
```

Comment:
This function repeats from the above, is there a way to refactor this to prevent duplication?

Reply:
Great suggestion! I have created one function now that is with the signature getHomeLocation(useAddress) where I pass in true if the request is from the address field and false if the request if from the browser location button. If useAddress is true, I get the value from the text field and use the getLocationFromAddress function otherwise I use the getLocationFromBrowser function.